### PR TITLE
CODENVY-1996: Improve checking sudo in rsync backup scripts

### LINKED ISSUE #1996: We should detect if sudo non-functional and don't use it for rsync scripts
Got this errors in logs:
<details>

```
2017-03-27 01:00:01,933[pScheduler-6216]  [ERROR] [DockerEnvironmentBackupManager 509]  - Error occurred during backup/restore of workspace 'workspace*******' on node 'node12.codenvy.io' : Warning: Permanently added '[node12.code
Authenticated to node12.codenvy.io ([172.30.12.112]:36526).
Transferred: sent 2336, received 2220 bytes, in 0.0 seconds
Bytes per second: sent 230170.4, received 218740.7
Warning: Permanently added '[node12.codenvy.io]:36526,[172.30.12.112]:36526' (ECDSA) to the list of known hosts.
Authenticated to node12.codenvy.io ([172.30.12.112]:36526).
sudo: no tty present and no askpass program specified
Transferred: sent 2396, received 2260 bytes, in 0.0 seconds
Bytes per second: sent 188812.6, received 178095.4
rsync: connection unexpectedly closed (0 bytes received so far) [Receiver]
rsync error: error in rsync protocol data stream (code 12) at io.c(226) [Receiver=3.1.2]
2017-03-27 01:00:01,935[pScheduler-6216]  [ERROR] [DockerEnvironmentBackupManager 356]  - Synchronization process failed. Exit code 12
java.io.IOException: Synchronization process failed. Exit code 12
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeCommand(DockerEnvironmentBackupManager.java:511) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeBackupScript(DockerEnvironmentBackupManager.java:347) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupInsideLock(DockerEnvironmentBackupManager.java:228) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupWorkspace(DockerEnvironmentBackupManager.java:127) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.WorkspaceFsBackupScheduler.lambda$scheduleBackup$0(WorkspaceFsBackupScheduler.java:106) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_111-internal]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
2017-03-27 01:00:01,936[pScheduler-6216]  [ERROR] [m.b.WorkspaceFsBackupScheduler 112]  - Backup of workspace workspace************** filesystem terminated on node12.codenvy.io node. Synchronization process failed. Exit code 12
org.eclipse.che.api.core.ServerException: Backup of workspace workspace********** filesystem terminated on node12.codenvy.io node. Synchronization process failed. Exit code 12
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeBackupScript(DockerEnvironmentBackupManager.java:359) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupInsideLock(DockerEnvironmentBackupManager.java:228) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupWorkspace(DockerEnvironmentBackupManager.java:127) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.WorkspaceFsBackupScheduler.lambda$scheduleBackup$0(WorkspaceFsBackupScheduler.java:106) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_111-internal]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
```
</details>

It looks like `rsync` doesn't like extra output that `sudo` produces, resulting in error code 12.
We need to fix it, either by not only checking that `sudo` command is present but testing if it functional, or by allocating pseudo-terminal by `ssh -t` if we know user password or by asking user for permission to modify sudo configuration to allow passwordless use of sudo.

Maybe it worth to consider using `dropbear` and/or `gosudo` instead of distributive specific versions, that may have altered configuration in images/recipes users trying to use.

**Reproduction Steps:**
Create new workspace from stack, I used python console simple.
Run workspace, open terminal window, and as root edit `/etc/sudoers` file:
just comment out last line that allows passwordless use of sudo:
```
#includedir /etc/sudoers.d
#%sudo ALL=(ALL) NOPASSWD: ALL
```
wait some time like 10 minutes and observe errors in log:
<details>

```
2017-03-28 13:30:38,956[kupScheduler-68]  [ERROR] [DockerEnvironmentBackupManager 509]  - Error occurred during backup/restore of workspace 'workspacebebdz483vl2b1hs9' on node 'node2.c5.codenvy-stg.com' : Warning: Permanently added '[node
2.c5.codenvy-stg.com]:40510,[172.30.12.84]:40510' (ECDSA) to the list of known hosts.
Warning: Permanently added '[node2.c5.codenvy-stg.com]:40510,[172.30.12.84]:40510' (ECDSA) to the list of known hosts.
sudo: no tty present and no askpass program specified
rsync: connection unexpectedly closed (0 bytes received so far) [Receiver]
rsync error: error in rsync protocol data stream (code 12) at io.c(226) [Receiver=3.1.2]
2017-03-28 13:30:38,957[kupScheduler-68]  [ERROR] [DockerEnvironmentBackupManager 356]  - Synchronization process failed. Exit code 12
java.io.IOException: Synchronization process failed. Exit code 12
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeCommand(DockerEnvironmentBackupManager.java:511) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeBackupScript(DockerEnvironmentBackupManager.java:347) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupInsideLock(DockerEnvironmentBackupManager.java:228) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupWorkspace(DockerEnvironmentBackupManager.java:127) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.WorkspaceFsBackupScheduler.lambda$scheduleBackup$0(WorkspaceFsBackupScheduler.java:106) [codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_111-internal]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
2017-03-28 13:30:38,958[kupScheduler-68]  [ERROR] [m.b.WorkspaceFsBackupScheduler 112]  - Backup of workspace workspacebebdz483vl2b1hs9 filesystem terminated on node2.c5.codenvy-stg.com node. Synchronization process failed. Exit code 12
org.eclipse.che.api.core.ServerException: Backup of workspace workspacebebdz483vl2b1hs9 filesystem terminated on node2.c5.codenvy-stg.com node. Synchronization process failed. Exit code 12
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.executeBackupScript(DockerEnvironmentBackupManager.java:359) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupInsideLock(DockerEnvironmentBackupManager.java:228) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.DockerEnvironmentBackupManager.backupWorkspace(DockerEnvironmentBackupManager.java:127) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at com.codenvy.machine.backup.WorkspaceFsBackupScheduler.lambda$scheduleBackup$0(WorkspaceFsBackupScheduler.java:106) ~[codenvy-machine-hosted-5.5.0.jar:5.5.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_111-internal]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_111-internal]
        at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_111-internal]
```
</details>

**Codenvy version:**    5.5.1-SaaS

### PR DESCRIPTION
**Pull request by garagatyi**
### What does this PR do?
Changes strategy of checking whether sudo should be used for rsync backuping.

### What issues does this PR fix or reference?
Fixes #1996 

#### Changelog
Improve checking sudo in rsync backup scripts

### CODE REVIEW

File: dockerfiles/init/modules/codenvy/templates/rsyncbase.sh.erb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
```
Comment (benoitf) [inline dockerfiles/init/modules/codenvy/templates/rsyncbase.sh.erb]: do we need copyright headers ?

Comment (garagatyi) [inline dockerfiles/init/modules/codenvy/templates/rsyncbase.sh.erb]: Added license header into all rsync scripts

File: dockerfiles/init/modules/codenvy/templates/rsyncbase.sh.erb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+SSH_OPTIONS+=" -o LogLevel=${SSH_LOG_LEVEL}"
+#######################
+
+# We need root permissions in target container to change ownership of files.
+# If user in that container has access to sudo we use it.
+# Otherwise we use rsync without sudo and if it is possible ownership will be changed.
+checkFile="/tmp/codenvy_sudo_test_$(date +%s)"
+createCheckFile="sh -c 'echo 1 > $checkFile' 2>/dev/null"
+checkSudo="$createCheckFile; hash sudo 2>/dev/null && sudo chown 1000 $checkFile 2>/dev/null"
+GET_RSYNC_COMMAND='if [ "$(id -u)" != "0" ]; then '$checkSudo' && echo "sudo rsync" && exit; fi; echo "rsync"'
```
Comment (skryzhny) [inline dockerfiles/init/modules/codenvy/templates/rsyncbase.sh.erb]: I'm proposing:
```
NEED_SUDO_COMMAND='[ "$(id -u)" != "0" ] && '$checkSudo' && echo " sudo"'
RSYNC_COMMAND="$(ssh ${SSH_OPTIONS} ${HOST} $(echo "$NEED_SUDO_COMMAND")) rsync"
```
 WDYT?

Comment (garagatyi) [inline dockerfiles/init/modules/codenvy/templates/rsyncbase.sh.erb]: `false && echo 1; echo 2` shows 2. It is what will happen in our case. It looks better but to make it work in a correct way I would need to introduce new constructions or commands. 
So I would leave it as is.

**Pull request merged**